ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java

```diff
@@ -1367,6 +1367,326 @@ public void onEvent(ReplicaId replicaId, Object e) {
+      assertTrue("Must encounter a cache miss", cacheMissLatch.await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertTrue("Must put metadata in cache", putCacheEntryLatch.await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+
+      // failed get
+      layout.getMockServers().forEach(mockServer -> mockServer.setServerErrorForAllRequests(serverErrorCode));
+      TestCallback<GetBlobResult> testCallback = new TestCallback<>();
+      Future<GetBlobResult> future = router.getBlob(blobId, new GetBlobOptionsBuilder().range(range).build(), testCallback, null);
+      assertFailureAndCheckErrorCode(future, testCallback, routerErrorCode);
+      assertTrue("Must encounter a cache hit", cacheHitLatch.await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      if (deleteCacheEntryOnFail) {
+        assertTrue("Must delete metadata in cache", deleteCacheEntryLatch.await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+        assertNull("Blob metadata must be absent in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+      } else {
+        assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+      }
+      layout.getMockServers().forEach(mockServer -> mockServer.resetServerErrors());
+      // There should be flag to disable this cache !
+      router.getNotFoundCache().invalidateAll();
+
+      // successful get, cache must repopulate
+      router.getBlob(blobId, new GetBlobOptionsBuilder().range(range).build()).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+    } finally {
+      router.close();
+    }
+  }
+
+  @Test
+  public void testRouterMetadataCacheBlobDoesNotExistOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Blob_Not_Found, RouterErrorCode.BlobDoesNotExist, true);
+  }
+  @Test
+  public void testRouterMetadataCacheBlobDeletedOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Blob_Deleted, RouterErrorCode.BlobDeleted, true);
+  }
+  @Test
+  public void testRouterMetadataCacheBlobExpiredOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Blob_Expired, RouterErrorCode.BlobExpired, true);
+  }
+  @Test
+  public void testRouterMetadataCacheUnexpectedInternalErrorOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Unknown_Error, RouterErrorCode.UnexpectedInternalError, false);
+  }
+
+  public void routerMetadataCacheErrorOnDeleteCompositeBlob(ServerErrorCode serverErrorCode, RouterErrorCode routerErrorCode) throws Exception {
```

Comment:
no need to create this function, we can just copy all the code to the test function below.

Reply:
yes but this would be needed if we want to test with different error combos. Right now there is just one, but I was hoping we'd add more in the future